### PR TITLE
Fix flaky instrumentation tests

### LINKED ISSUE #4429: Refactor incorrect `CoreRerouteTest` and `RouteAlternativesTest`
An issue caught by an Instrumentation test found while working on https://github.com/mapbox/mapbox-navigation-android/pull/4419 https://app.circleci.com/pipelines/github/mapbox/mapbox-navigation-android/12041/workflows/6cc8f997-07d1-49f4-a898-9243e7ace970/jobs/48760 https://console.firebase.google.com/u/0/project/mapbox-navigation-android/testlab/histories/bh.54eb2dd57e959221/matrices/5010895745904766073/executions/bs.1e3c7202ff856a89/testcases/2/errors

```
java.lang.AssertionError
at org.junit.Assert.fail(Assert.java:87)
at org.junit.Assert.assertTrue(Assert.java:42)
at org.junit.Assert.assertTrue(Assert.java:53)
at com.mapbox.navigation.instrumentation_tests.core.RouteAlternativesTest.expect_faster_route_alternatives(RouteAlternativesTest.kt:79)
at java.lang.reflect.Method.invoke(Native Method)
at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
at androidx.test.internal.runner.junit4.statement.RunBefores.evaluate(RunBefores.java:80)
at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
at androidx.test.rule.GrantPermissionRule$RequestPermissionStatement.evaluate(GrantPermissionRule.java:134)
at androidx.test.rule.ActivityTestRule$ActivityStatement.evaluate(ActivityTestRule.java:531)
at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
at com.schibsted.spain.barista.rule.cleardata.ClearDatabaseRule$apply$1.evaluate(ClearDatabaseRule.kt:30)
at com.schibsted.spain.barista.rule.cleardata.ClearPreferencesRule$1.evaluate(ClearPreferencesRule.java:26)
at com.schibsted.spain.barista.rule.cleardata.ClearFilesRule$apply$1.evaluate(ClearFilesRule.kt:25)
at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
at org.junit.rules.RunRules.evaluate(RunRules.java:20)
at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
at org.junit.runners.Suite.runChild(Suite.java:128)
at org.junit.runners.Suite.runChild(Suite.java:27)
at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
at androidx.test.internal.runner.TestExecutor.execute(TestExecutor.java:56)
at androidx.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:395)
at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:2074)
```

cc @RingerJK @kmadsen

### PR DESCRIPTION
**Pull request by Guardiola31337**
### Description

Fixes flaky instrumentation tests
Fixes #4429 

Opening this as a `Draft` for debugging purposes.

- attempt to record history file of a failing test

### CODE REVIEW

**Comment by Guardiola31337** [comment]
Making good progress here. I was able to upload the history files generated from the tests into the Google Cloud test bucket 🚀 

Screen Shot 2021-08-20 at 3 38 07 PM

party

**Comment by Guardiola31337** [comment]
After https://github.com/mapbox/mapbox-navigation-android/pull/4919 this is now 💚 

Thanks @mskurydin @SiarheiFedartsou for looking into this 🙇‍♂️

**Comment by Guardiola31337** [comment]
This is ready for review 💃

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt [hunk 1]
```diff
@@ -123,36 +138,83 @@ class CoreRerouteTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.jav
                         reasons: List<RouterFailure>,
                         routeOptions: RouteOptions
                     ) {
-                        // no impl
+                        Log.e("DEBUG", "onFailure reasons=$reasons")
```
Comment (kmadsen) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt:141]: nit: should we be using `LoggerProvider.logger`?

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt [hunk 2]
```diff
@@ -123,36 +138,83 @@ class CoreRerouteTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.jav
...
                     }
                 }
             )
         }
 
         // wait for tracking to start
-        Espresso.onIdle()
+        try {
+            Espresso.onIdle()
+        } catch (ex: RuntimeException) {
```
Comment (kmadsen) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt:157]: Which `RuntimeException` is being thrown here? Curious why we expect it 🤔 . Is this to help us catch issues when they arise?

Comment (kmadsen) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt:157]: We could add this check to the `MapboxHistoryTestRule`

for example. add a function

``` kotlin
mapboxHistoryTestRule.stopRecordingOnCrash {
    Espresso.onIdle()
}
```

``` kotlin
MapboxHistoryTestRule {
    fun stopRecordingOnCrash(runner: () -> Unit) {
        try {
            runner()
        } catch (t: Throwable) {
            val countDownLatch = CountDownLatch(1)
            mapboxNavigation.historyRecorder.stopRecording {
                Log.e("DEBUG", "no location tracking history path=$it")
                countDownLatch.countDown()
            }
            countDownLatch.await()
        }
    }
}

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt [hunk 3]
```diff
@@ -40,19 +43,25 @@ class RouteAlternativesTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::cla
...
     @Before
     fun setup() {
-        val routeAlternativesOptions = RouteAlternativesOptions.Builder()
-            .intervalMillis(TimeUnit.SECONDS.toMillis(10))
-            .build()
-        mapboxNavigation = MapboxNavigationProvider.create(
-            NavigationOptions.Builder(activity)
-                .accessToken(getMapboxAccessTokenFromResources(activity))
-                .routeAlternativesOptions(routeAlternativesOptions)
+        runOnMainSync {
```
Comment (kmadsen) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:53]: is this the source of the issue? not running the MapboxNavigationProvider.create on the main sync?

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt (hunk 2 above)
Comment (Guardiola31337) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt:157]: > Which `RuntimeException` is being thrown here? Curious why we expect it 🤔 . Is this to help us catch issues when they arise?

```
java.lang.RuntimeException: java.util.concurrent.ExecutionException: androidx.test.espresso.IdlingResourceTimeoutException: Wait for [RouteProgressStateIdlingResource] to become idle timed out
	at androidx.test.espresso.Espresso.onIdle(Espresso.java:295)
	at androidx.test.espresso.Espresso.onIdle(Espresso.java:314)
	at com.mapbox.navigation.instrumentation_tests.core.MapboxHistoryTest.verify_history_files_are_recorded_and_readable(MapboxHistoryTest.kt:142)
	at java.lang.reflect.Method.invoke(Native Method)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at androidx.test.internal.runner.junit4.statement.RunBefores.evaluate(RunBefores.java:80)
	at androidx.test.internal.runner.junit4.statement.RunAfters.evaluate(RunAfters.java:61)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at androidx.test.rule.GrantPermissionRule$RequestPermissionStatement.evaluate(GrantPermissionRule.java:134)
	at androidx.test.rule.ActivityTestRule$ActivityStatement.evaluate(ActivityTestRule.java:531)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at com.schibsted.spain.barista.rule.cleardata.ClearDatabaseRule$apply$1.evaluate(ClearDatabaseRule.kt:30)
	at com.schibsted.spain.barista.rule.cleardata.ClearPreferencesRule$1.evaluate(ClearPreferencesRule.java:26)
	at com.schibsted.spain.barista.rule.cleardata.ClearFilesRule$apply$1.evaluate(ClearFilesRule.kt:25)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at androidx.test.internal.runner.TestExecutor.execute(TestExecutor.java:56)
	at androidx.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:395)
	at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:1879)
Caused by: java.util.concurrent.ExecutionException: androidx.test.espresso.IdlingResourceTimeoutException: Wait for [RouteProgressStateIdlingResource] to become idle timed out
	at java.util.concurrent.FutureTask.report(FutureTask.java:94)
	at java.util.concurrent.FutureTask.get(FutureTask.java:164)
	at androidx.test.espresso.Espresso.onIdle(Espresso.java:287)
	... 46 more
Caused by: androidx.test.espresso.IdlingResourceTimeoutException: Wait for [RouteProgressStateIdlingResource] to become idle timed out
	at androidx.test.espresso.IdlingPolicy.handleTimeout(IdlingPolicy.java:61)
	at androidx.test.espresso.base.UiControllerImpl$5.resourcesHaveTimedOut(UiControllerImpl.java:428)
	at androidx.test.espresso.base.IdlingResourceRegistry$Dispatcher.handleTimeout(IdlingResourceRegistry.java:478)
	at androidx.test.espresso.base.IdlingResourceRegistry$Dispatcher.handleMessage(IdlingResourceRegistry.java:411)
	at android.os.Handler.dispatchMessage(Handler.java:98)
	at androidx.test.espresso.base.Interrogator.loopAndInterrogate(Interrogator.java:148)
	at androidx.test.espresso.base.UiControllerImpl.loopUntil(UiControllerImpl.java:519)
	at androidx.test.espresso.base.UiControllerImpl.loopMainThreadUntilIdle(UiControllerImpl.java:441)
	at androidx.test.espresso.Espresso$1.run(Espresso.java:277)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:423)
	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
	at android.os.Handler.handleCallback(Handler.java:739)
	at android.os.Handler.dispatchMessage(Handler.java:95)
	at android.os.Looper.loop(Looper.java:148)
	at android.app.ActivityThread.main(ActivityThread.java:5417)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
```

And yeah the idea is to capture and retrieve the history files generated up that point, so we can troubleshoot any potential regressions easily. We should add actually history files support to all of our Instrumentation tests. Will cut a follow up ticket.

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt (hunk 3 above)
Comment (Guardiola31337) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:53]: Yeah one of them .On instrumentation tests we need to run the `MapboxNavigation` creation on the main thread, if not the navigator observer is not set and all the interactions with the `Navigator` are lost (NN added that "recently"). The were a couple more. 1, had to relax expected coordinates to handle MockWebServer re-routes (as we don't know exactly when NN is going to emit the `OFF_ROUTE`) and 2, the NN 32-bit / fallback mode 💥

**Review by kmadsen — APPROVED** [review]
A couple refactor suggestions, but merge when you're ready! 🚀 

Nice work!!

**Comment by Guardiola31337** [comment]
@kmadsen addressed the feedback, this is ready for another round of 👀

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt (hunk 2 above)
Comment (Guardiola31337) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt]: > Will cut a follow up ticket.

https://github.com/mapbox/mapbox-navigation-android/issues/4927

**Pull request merged**